UnityProject/Assets/Scripts/Core/Chat/Chat.cs

```diff
@@ -121,9 +121,49 @@ public static void AddChatMsgToChat(ConnectedPlayer sentByPlayer, string message
 			speaker = (player == null) ? sentByPlayer.Username : player.playerName,
 			position = (player == null) ? TransformState.HiddenPos : player.PlayerChatLocation.AssumedWorldPosServer(),
 			channels = channels,
-			originator = (player == null) ? sentByPlayer.GameObject : player.PlayerChatLocation
+			originator = (player == null) ? sentByPlayer.GameObject : player.PlayerChatLocation,
+			VoiceLevel = loudness
 		};
 
+		bool IsOnCorrectChannels()
+		{
+			if (channels.HasFlag(ChatChannel.Common) ||
+			    channels.HasFlag(ChatChannel.Command) || channels.HasFlag(ChatChannel.Security)
+			    || channels.HasFlag(ChatChannel.Engineering) || channels.HasFlag(ChatChannel.Medical)
+			    || channels.HasFlag(ChatChannel.Science)
+			    || channels.HasFlag(ChatChannel.Syndicate) || channels.HasFlag(ChatChannel.Supply))
+			{
+				return true;
+			}
+			return false;
+		}
+
+		//Check if is not a ghost/spectator and the player has an inventory.
+		void CheckVoiceLevel(PlayerScript script)
+		{
+			if (!script.mind.body.IsDeadOrGhost && script.DynamicItemStorage != null)
+			{
+				foreach (ItemSlot slot in script.DynamicItemStorage.GetNamedItemSlots(NamedSlot.ear))
+				{
+					Headset headset = slot.Item?.gameObject.GetComponent<Headset>();
+					if (headset != null)
+					{
+						if (headset.LoudSpeakOn && IsOnCorrectChannels())
+						{
+							chatEvent.VoiceLevel = headset.LoudspeakLevel;
+						}
+					}
+				}
+			}
+		}
```

Comment:
Could you move this into its own function not a local one, its not needed inside and bloats it

Reply:
Means you can unnest it all too

Reply:
Makes no sense for me to turn this into a whole proper function when it's never going to be used else where

Reply:
Its just the function is now 134 lines long and needs reorganising

Reply:
done